api/v0/ingest/schema/types.go

```diff
@@ -0,0 +1,178 @@
+package schema
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/multiformats/go-multihash"
+)
+
+type (
+	Advertisement struct {
+		PreviousID *ipld.Link
+		Provider   string
+		Addresses  []string
+		Signature  []byte
+		Entries    ipld.Link
+		ContextID  []byte
+		Metadata   []byte
+		IsRm       bool
+	}
+	EntryChunk struct {
+		Entries []multihash.Multihash
+		Next    *ipld.Link
+	}
+
+	// TODO: Propose this to go-ipld-prime as a way to reduce boilerplate code when working with linksystem?
+	//       Loader interface {
+	//       	FromNode(ipld.Node) error
+	//       	LoadFrom(ipld.LinkContext, ipld.LinkSystem, ipld.Link) error
+	//       }
+	//       Storer interface {
+	//       	ToNode() (ipld.Node, error)
+	//       	StoreTo(ipld.LinkContext, ipld.LinkSystem) (ipld.Link, error)
+	//       }
+)
+
+// ToNode converts this advertisement to its representation as an IPLD typed node.
+//
+// See: bindnode.Wrap.
+func (a Advertisement) ToNode() (n ipld.Node, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = fmt.Errorf("unknown panic: %v", r)
+			}
+		}
+	}()
```

Comment:
I would hope that you can remove this recover at some point, once your code and bindnode is more stable :) perhaps add a TODO for that?

it should only panic if there's a bug in bindnode, or if you put values inside `a` which are incompatible with the schema, e.g. an union with no value set (which is disallowed).